### ISSUE TITLE
Use of Performance cores on macOS Apple Silicon (Mx series CPUs)

### ISSUE THREAD

**Issue by siraaris** [HEnquist/camilladsp#357]
It would be good to have a mechanism for camilladsp on macOS to utilise Performance cores on the M series Apple Silicon CPU's.

Under some loads, eg high channel count, high sample rate and reasonably large FIR filters, the Efficiency cores may not be suitable, e.g. I've observed glitches, high load reported from camilladsp when running on Efficiency cores that seem to go away when the OS promotes the process to Performance cores (investigation is ongoing!).

I think thread priority must be set in code, as there's no option the I'm aware of for this to be set in user land.

**Comment by siraaris** [comment]
To reduce load I created a set of unified FIR filters to get my config down to the most efficient possible, so now I ran single FIR filters per channel incorporating XO, DRC and other DSP etc. 

Glitches (restarts) are observed even when CamillaDSP is quiescent. The buffer levels are also stable. 

It would seem to suggest that the OS, BlackHole or CoreAudio is doing something here.

Any suggestions where to look? I'm running out of ideas (until you get time to look at thread prioritisation :), no drama just sharing what I've noticed.

**Comment by siraaris** [comment]
Hmmm. Looks like Restarts/Buffer Levels occur hourly. Screams cronjob, will investigate, or could be when macOS maintains processes.

<img width="1262" alt="Screenshot 2024-09-04 at 12 55 55 AM" src="https://github.com/user-attachments/assets/6586c38d-dd1b-4228-a9e0-0b03df8d67d4">

**Comment by HEnquist** [comment]
I have made some changes to improve the buffer level measurement, not sure if it helps for this issue or not, but you could the the latest next30 to see if it helps. I also added a script for plotting the processing load and buffer level while running: https://github.com/HEnquist/camilladsp/blob/next30/testscripts/log_load_and_level.py
I have let it run for a couple of hours on my M1 Air, hoping to get results similar to yours. But it behaves well without buffer underruns or spikes in the level.

**Comment by siraaris** [comment]
That's a useful utility :)

I think there are a number of factors in my case. My audio devices are AVB based, that are probably uncommon for most users of CamillaDSP.

For example, the Playback device is presented to CoreAudio from a USB connected device (RME Digitface AVB) that bridges into network transport to connect to the DAC (RME M32 DA Pro, a 32 ch DAC), managed by RME's AVB Controller.

This should be transparent to CamillaDSP if  everything else is well behaved.

Things I've noticed that result in CamillaDSP glitches:

1. RME have moved the driver for the Digiface AVB from a Kernel Extension to a System Extension (DriverKit), which moves runtime from kernel space to user space. I don't think this is the direct cause of any of my observed issues - but it's worth noting.
2. RME have also moved the ethernet driver for the Digiface AVB (netifc) into user space. Again, I don't think directly an issue when netifc is behaving well (see next point), other than another part of the process that's changed over time with Apple deprecating kernel extensions over time.
3. The RME netifc process continually segmentation fault crashes. I've filed a support request with RME. From (admittedly limited) data over the last few days or so. I think there is a correlation between observed (audible) disturbances in CamillaDSP and when netifc crashes and restarts.
4. The Mx series Apple Silicon CPU's have a mixture of Efficiency and Performance cores. The M1 for example has 4 of each. Based on (once again, limited) data over the last week or so, I think there may be an effect on CamillaDSP when the running process is moved to/from different type cores - but I need to look further, and maybe what I'm observing is directly because of core swapping.

I still think it's worth implementing thread prioritisation / core affinity or similar for macOS.

Because of the above I've spent some time tonight getting an AVB configuration setup only using the RME DAC and Apple built-in AVB Controller (avbutil), see how that goes.

**Comment by siraaris** [comment]
Removing the RME devices as above, the behaviour overnight still suggests an hourly incident.
<img width="1294" alt="Screenshot 2024-09-05 at 3 18 28 PM" src="https://github.com/user-attachments/assets/fcc37583-449b-471b-a7e8-073df07764dd">

The three blips in the centre (around 20000) are at 7:27, 8:27, and 9:27, followed by a long stretch of stability, until I removed in, restarted looking at this on console.

At these times there are lots of:

2024-09-05 07:27:02.757878 INFO [src/coreaudiodevice.rs:446] Restarting playback after buffer underrun
2024-09-05 07:27:03.951268 WARN [src/coreaudiodevice.rs:455] Playback interrupted, no data available
...
2024-09-05 08:27:19.185577 WARN [src/coreaudiodevice.rs:455] Playback interrupted, no data available
2024-09-05 08:27:19.189584 INFO [src/coreaudiodevice.rs:446] Restarting playback after buffer underrun
...
2024-09-05 09:27:19.021137 INFO [src/coreaudiodevice.rs:446] Restarting playback after buffer underrun
2024-09-05 09:27:20.214530 WARN [src/coreaudiodevice.rs:455] Playback interrupted, no data available
...

**Comment by siraaris** [comment]
Oh my dear lord. 

https://eclecticlight.co/2023/01/21/how-macos-schedules-background-activities/

**Comment by siraaris** [comment]
A bit of progress. I removed the Presonus AVB switch, leaving only the Mac mini connected directly to the DAC via ethernet.

Lo-and behold, things seem to settle down, see screen grab.

<img width="1857" alt="Screenshot 2024-09-06 at 3 35 06 AM" src="https://github.com/user-attachments/assets/4c6ff71c-bdf4-4f05-8910-8f2fce56dd14">

And the RME networking (netifc) is not crashing.

Will let it run this weekend, and provide an update.

**Comment by HEnquist** [comment]
It looks like macOS offers very limited control over thread priorities. There is a concept of quality of service, but not sure if it's applicable here. Then there is audio workgroups, as described here: https://www.bluecataudio.com/Blog/announcements/realtime-audio-multicore-issues-for-apple-silicon-end-of-the-story/
That looks interesting, but much more difficult to use than just adjusting some thread priorities. Not sure how feasible it is to use it in camilladsp.

**Comment by HEnquist** [comment]
Found this that I will try: https://crates.io/crates/audio_thread_priority

**Comment by siraaris** [comment]
<img width="1876" alt="Screenshot 2024-09-06 at 9 19 45 AM" src="https://github.com/user-attachments/assets/418fe012-2b60-4854-8c92-d795adedb89f">

Behaviour overnight. Just FYI.

**Comment by siraaris** [comment]
> It looks like macOS offers very limited control over thread priorities. There is a concept of quality of service, but not sure if it's applicable here. Then there is audio workgroups, as described here: https://www.bluecataudio.com/Blog/announcements/realtime-audio-multicore-issues-for-apple-silicon-end-of-the-story/ That looks interesting, but much more difficult to use than just adjusting some thread priorities. Not sure how feasible it is to use it in camilladsp.

Seems clear that using audio workgroups addresses the issue. It's interesting that Intel CPU's are also introducing CPU's with varying core performance specifications; maybe this issue will start to exhibit on non-macOS as well.

What's not clear is detail on the "hack" that BlueCatAudio refer to. Maybe it's what audio_thread_priority Rust crate utilises?

**Comment by HEnquist** [comment]
> What's not clear is detail on the "hack" that BlueCatAudio refer to. Maybe it's what audio_thread_priority Rust crate utilises?

The audio_thread_priority crate doesn't seem to be using audio workgroups, so probably not. But we don't know what that hack is (only that it's supposedly obvious if you look somewhere in the apple open source code 😒) so this is only guess.

I added audio_thread_priority to in the processing thread and the CoreAudio capture and playback threads in branch `audio_thread_prio`. Can you try it on your system?

**Comment by siraaris** [comment]
I am running the audio_thread_prio branch now - will report back in a few hours, which should be enough to see how it goes.

**Comment by siraaris** [comment]
Initial observation - buffers are very stable. I've tried spiking the CPU with various actions, which previously would trigger buffers to build and crash - so far looking good.

![Uploading Screenshot 2024-09-07 at 10.10.04 PM.png…]()

**Comment by siraaris** [comment]
Initial observation - buffers are very stable. I've tried spiking the CPU with various actions, which previously would trigger buffers to build and crash - so far looking good.

<img width="1920" alt="Screenshot 2024-09-07 at 10 10 04 PM" src="https://github.com/user-attachments/assets/e174947c-d04e-43b4-be78-f7a00e7d5a3b">

**Comment by siraaris** [comment]
Well, the thread priority change has resulted in super stability for camilladsp. No dropouts/restarts, stable buffers, really nice. There's a bit of lagginess on the UI when I remote in, but the Mac mini is dedicated for camilladsp so that's a small price to pay for audio stability and performance.

I'll keep it running, and report back. If there's anything you need me to look at specifically, just shout out.

<img width="1640" alt="Screenshot 2024-09-08 at 1 58 19 AM" src="https://github.com/user-attachments/assets/be489dad-4879-4719-a2f8-66ebdffb82ac">

**Comment by HEnquist** [comment]
Looks great so far! I don't have any specific things I want tested, just curious about how it behaves when kept running for a while.

**Comment by siraaris** [comment]
I think you can probably treat this as "done". Not really sure that Performance cores are actually used, but regardless makes no difference - the result is stable and performant behaviour.

<img width="1492" alt="Screenshot 2024-09-08 at 3 36 36 PM" src="https://github.com/user-attachments/assets/bc380b2c-87e8-43c8-935b-437e1486037a">

**Comment by HEnquist** [comment]
Would you be interested in trying multithreaded processing? The branch "with_rayon" supports splitting filter tasks among several threads. This is enabled via a new optional boolean `multithreaded` in the devices section of the config (that defaults to off).

It hasn't gotten much testing, so please start with amplifiers powered off :)

The idea is that between mixers and processors, each channel can be filtered independently from the others. So it collects the filters to apply to each channel, and then uses the really smart rayon library to process the channels in parallel in a set of worker threads.
It needs quite heavy filter tasks to actually help, with too "easy" filters the overhead of passing things back and forth between threads gets larger than the actual processing time. I think your config could potentially benefit.

**Comment by siraaris** [comment]
I tried the with_rayon branch.

Seems to run ok, but now get underruns - I think because the threads you create per channel need also to be real-time?

For example - with multithreaded: false, I can stop/start Safari (ie cause CPU spikes), and camilladsp is unaffected.

With multithreaded: true, stop/start Safari causes camilladsp to hiccup.

**Comment by HEnquist** [comment]
> now get underruns - I think because the threads you create per channel need also to be real-time?

Yes those also need to have their priority raised, just didn't get to that yet. But did you see any change to the processing load? Hopefully it should be lower.

**Comment by siraaris** [comment]
I'll run for a while with graphing and let you know.

**Comment by siraaris** [comment]
The load is lower yes.

<img width="1753" alt="Screenshot 2024-09-10 at 3 51 48 PM" src="https://github.com/user-attachments/assets/29358522-b3d9-4eac-b508-09bb98503dcb">

**Comment by siraaris** [comment]
Longer timeframe. You can see the spike at the end when I remote in a screen capture.

<img width="1753" alt="Screenshot 2024-09-10 at 4 22 17 PM" src="https://github.com/user-attachments/assets/6de91edb-7efe-49f5-bc5b-1a9653e9e03c">

**Comment by HEnquist** [comment]
Ok! Thanks for testing. I would expect the threading to make it more sensitive to interference from other loads.   Raising priorities should help, but there may still be delays when waking up the worker threads, and when they notify the main processing thread that they are finished.

**Comment by siraaris** [comment]
I'll happily test! I think it's worth pursuing and having the support there?

**Comment by HEnquist** [comment]
The `with_rayon` branch is updated. Now it raises the priority of the workers, and the number of workers can be set by the `worker_threads` parameter in devices. Leave it out or set to 0 to let rayon decide, which becomes one thread per hardware thread of the machine. On the Windows laptop I'm using at the moment (12-core Snapdragon X Elite cpu), anything above 4 threads gives the same processing load.

**Comment by siraaris** [comment]
Initial observation is that setting worker_threads manually is required, as the default number (when I set to 0) may be too high, and deterimental.

On my Mac mini, I've set worker_threads to 4 and that seems to be ok.

Will leave it running for a while and report back.

**Comment by siraaris** [comment]
Final check in, only stable at 192khz, 32 ch with 2 threads. 

But with that it's rock solid.

**Comment by siraaris** [comment]
Without creating a new issue/suggestion, I've been experimenting - on Linux - on pinning camilladsp to specific CPU's as I'm still experiencing xrun issues.

Summary of steps I've taken:

Disabled Hyperthreading in BIOS
Installed RT kernel (6.10.11-rt-amd64)
tuned-adm latency-performance
I have an 8-core i7, so:
Boot kernel with isolcpus=0,1,2,3,4,5 (on Debian, edit /etc/default/grub and run update-grub)
Set CamillaDSP config:
  multithreaded: true
  worker_threads: 4
Start camilladsp with taskset --cpu-list 0-5

**Comment by HEnquist** [comment]
Does pinning make any difference?

**Comment by siraaris** [comment]
I'm not sure yet, but for the above configuration there are still underruns reported:

aris@controller:~$ tail -f proj/log/camilladsp.log 
2024-10-06 16:37:35.801437 INFO  [src/bin.rs:781] CamillaDSP version 3.0.0
2024-10-06 16:37:35.801443 INFO  [src/bin.rs:782] Running on linux, x86_64
2024-10-06 16:37:35.924265 INFO  [src/alsadevice.rs:789] Capture device supports rate adjust
2024-10-06 16:37:36.036685 INFO  [src/alsadevice.rs:117] PB: Starting playback from Prepared state
2024-10-06 17:21:36.039657 WARN  [src/alsadevice.rs:113] PB: Prepare playback after buffer underrun
2024-10-06 17:21:51.833882 WARN  [src/alsadevice.rs:113] PB: Prepare playback after buffer underrun
2024-10-06 18:31:57.670345 WARN  [src/alsadevice.rs:113] PB: Prepare playback after buffer underrun
2024-10-06 18:45:02.588972 WARN  [src/alsadevice.rs:113] PB: Prepare playback after buffer underrun

I don't think it's a system issue per-se, as the same filters, and pipeline setup on Brutefir doesn't exhibit xruns (well, at least for 12 hours at a stretch). For the "same" configuration, CamillaDSP run's every hour or two.

**Comment by siraaris** [comment]
Cognisant that this isn't macOS related, do you want me to create a new issue - we can probably close this one, as macOS on M1 Mac with the changes you introduced were solid (provided that worker_threads wasn't too high.

**Comment by HEnquist** [comment]
Yeas that fits better in a new issue. Please attach the config file, and the output of aplay -l and arecord -l.
